### PR TITLE
Ui fixed fees redesign

### PR DESCRIPTION
**Pull request by cleargif**
#### What
Apply new design to fixed fees page

#### Ticket

[Pivotal](https://www.pivotaltracker.com/story/show/157116534)

#### Why
Simon asked nicely

#### How
CSS:
As we apply updated designs to the pages we will migrate the pages from V1 css files to V2. The migration is to simplify the CSS and pay off some tech dept. Some linting also apllied to the scss

note: during this migration their will be instances of duplication across both V1 and V2. This is as a result of shared views / form builders.

JS:
- minor config updates on Error handling module to work with updated semantics
- Refactor to remove nasty looking ternary statement

Ruby:
- `adp_text_field` updated semantics,  some styles and the specs
- `claims_helper` updated with css
- `base_fee` and `basic_fee` models updated
- Features updated as result of semantic changes

Views:
- implemented fixed fee design
- `dates_attended_fields_condensed` created to aid in migrating pages with shared templates
- using the `v2_stylesheets` value to control the loading of css and `basic` & `fixed` fee templates

GEMS:
- `gov_uk_date_fields` updated

Translations:
- updated to account for new structures in the views

### CODE REVIEW

File: app/views/external_users/claims/_date_attended_fields_condenced.html.haml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+%fieldset
```
Comment (jsugarman) [inline app/views/external_users/claims/_date_attended_fields_condenced.html.haml]: **minor - non-blocker** file name `..condenced --> condensed`

File: app/views/external_users/claims/basic_fees/_basic_fee_fields.html.haml [hunk 2]
```diff
@@ -13,8 +13,8 @@
 
         %span.no-dates.form-hint
           = t('.no_dates')
-        = link_to_add_association t('.add_date_attended'), f, :dates_attended, class: '', data: {'association-insertion-method' => 'prepend', 'association-insertion-traversal' => 'next', 'association-insertion-node' => 'div'}
+        = link_to_add_association t('.add_date_attended'), f, :dates_attended, class: '', partial: 'date_attended_fields_condenced', data: {'association-insertion-method' => 'prepend', 'association-insertion-traversal' => 'next', 'association-insertion-node' => 'div'}
```
Comment (jsugarman) [inline app/views/external_users/claims/basic_fees/_basic_fee_fields.html.haml]: as above `..condenced --> condensed`

File: app/views/external_users/claims/basic_fees/_basic_fee_fields.html.haml [hunk 3]
```diff
@@ -13,8 +13,8 @@
...
 
         .grid-row.dates-wrapper
           = f.fields_for :dates_attended do |date_attended|
-            = render 'date_attended_fields', { f: date_attended, submodel_count: date_attended.index+1, parent_model_prefix: "basic_fee_#{@basic_fee_count}" }
+            = render 'date_attended_fields_condenced', { f: date_attended, submodel_count: date_attended.index+1,  parent_model_prefix: "basic_fee_#{@basic_fee_count}" }
```
Comment (jsugarman) [inline app/views/external_users/claims/basic_fees/_basic_fee_fields.html.haml]: as above `..condenced --> condensed`

File: app/views/external_users/claims/misc_fees/_misc_fee_fields.html.haml [hunk 4]
```diff
@@ -37,10 +37,10 @@
       %label.form-label Dates
       %span.no-dates.form-hint
         = t('.no_dates')
-      = link_to_add_association t('.add_date_attended'), f, :dates_attended, class: '', data: {'association-insertion-method' => 'prepend','association-insertion-traversal' => 'next', 'association-insertion-node' => 'div'}
+      = link_to_add_association t('.add_date_attended'), f, :dates_attended, class: '', partial: 'date_attended_fields_condenced', data: {'association-insertion-method' => 'prepend','association-insertion-traversal' => 'next', 'association-insertion-node' => 'div'}
```
Comment (jsugarman) [inline app/views/external_users/claims/misc_fees/_misc_fee_fields.html.haml]: as above `..condenced --> condensed`

File: app/views/external_users/claims/misc_fees/_misc_fee_fields.html.haml [hunk 5]
```diff
@@ -37,10 +37,10 @@
...
 
       .grid-row.dates-wrapper
         = f.fields_for :dates_attended do |date_attended|
```
Comment (jsugarman) [inline app/views/external_users/claims/misc_fees/_misc_fee_fields.html.haml]: as above `..condenced --> condensed`

File: app/views/external_users/claims/misc_fees/_misc_fee_fields.html.haml [hunk 6]
```diff
@@ -37,10 +37,10 @@
...
-          = render 'date_attended_fields', f: date_attended, submodel_count: date_attended.index+1, parent_model_prefix: "misc_fee_#{@misc_fee_count}"
+          = render 'date_attended_fields_condenced', f: date_attended, submodel_count: date_attended.index+1, parent_model_prefix: "misc_fee_#{@misc_fee_count}"
```
Comment (jsugarman) [inline app/views/external_users/claims/misc_fees/_misc_fee_fields.html.haml]: as above `..condenced --> condensed`

File: config/locales/en.yml [hunk 7]
```diff
@@ -680,7 +680,10 @@ en:
           at any time in the <a href="%{external_users_root_url}">Your claims</a> view.</p>
         will_be_processed: Your claim will be processed shortly
       date_attended_fields:
-        date_from: 'Date attended (from)'
+        date: 'Date'
+        remove: 'Remove date attended'
+      date_attended_fields_condenced:
```
Comment (jsugarman) [inline config/locales/en.yml]: as above `..condenced --> condensed`

File: Gemfile [hunk 8]
```diff
@@ -13,7 +13,7 @@ gem 'devise',                 '~> 4.4.3'
 gem 'dropzonejs-rails',       '~> 0.8.2'
 gem 'factory_bot_rails',      '~> 4.8.2'
 gem 'faker',                  '~> 1.8.4'
-gem 'gov_uk_date_fields',     '~> 2.1.1'
+gem 'gov_uk_date_fields',     '~> 3.0.0'
```
Comment (jsugarman) [inline Gemfile:16]: 👍

File: app/assets/javascripts/application.js [hunk 9]
```diff
@@ -88,7 +88,7 @@ if (!String.prototype.supplant) {
 
     moj.Modules.Autocomplete.typeaheadKickoff(insertedSelect);
     moj.Modules.Autocomplete.typeaheadBindEvents(typeaheadWrapper);
-    moj.Modules.FeeFieldsDisplay.addFeeChangeEvent($insertedItem.find('.fx-fee-group'))
+    moj.Modules.FeeFieldsDisplay.addFeeChangeEvent(insertedItem)
```
Comment (jsugarman) [inline app/assets/javascripts/application.js:91]: unsure if this should be `$insertedItem` or is using the arg `insertedItem` passed directly ok?

File: app/assets/stylesheets/v1/application.css.scss [hunk 10]
```diff
@@ -203,3 +203,22 @@ form{
 
 .highlight { background-color: #ffffca }
 
+
+.field_with_errors, .dropdown_field_with_errors{
+  // border-left: 3px solid $error-colour;
```
Comment (lostie) [inline app/assets/stylesheets/v1/application.css.scss]: ✂️ ✂️

File: app/assets/stylesheets/v2/_autocomplete.scss [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+//autocomplete styling
+.tt-menu {
+  background-color: #fff;
+  border: 1px solid #ccc;
+  width: 100%;
+  max-height: 220px;
+  overflow-y: scroll;
+}
+
+
```
Comment (lostie) [inline app/assets/stylesheets/v2/_autocomplete.scss]: ✂️ ✂️

File: app/assets/stylesheets/v2/_forms.scss [hunk 12]
```diff
@@ -0,0 +1,27 @@
+.form-control:disabled {
```
Comment (lostie) [inline app/assets/stylesheets/v2/_forms.scss]: > DEPRECATION WARNING on line 127 of app/assets/stylesheets/v1/moj/_forms.scss:
> Extending a compound selector, .form-control:disabled, is deprecated and will not be supported in a future release.
> See https://github.com/sass/sass/issues/1599 for details.

Should we take any action on this? Ideally we should not add new styles that are already considered deprecated.

File: app/form_builders/adp_text_field.rb [hunk 13]
```diff
@@ -109,13 +109,6 @@ def generate_form_field_id_from_object_name
     else
       "#{@form.object_name}_#{@method}"
     end
-
-    # if @form.object_name.is_a?(Symbol)
```
Comment (lostie) [inline app/form_builders/adp_text_field.rb:113]: 🏆

File: app/views/external_users/claims/_date_attended_fields.html.haml [hunk 14]
```diff
@@ -1,23 +1,10 @@
...
+  - if defined?(parent_model_prefix) && defined?(submodel_count)
+    - error_prefix = "#{parent_model_prefix}_date_attended_#{submodel_count}"
+  - else
+    - error_prefix = "date_attended_0"
 
-    - if defined?(parent_model_prefix) && defined?(submodel_count)
-      - error_prefix = "#{parent_model_prefix}_date_attended_#{submodel_count}"
-    - else
-      - error_prefix = "date_attended_0"
+  = link_to_remove_association "Remove", f, {class: "date-action"}
```
Comment (lostie) [inline app/views/external_users/claims/_date_attended_fields.html.haml]: **[not a blocker]** There's some locales defined for buttons. Would be nice to start moving some of this to it.

File: app/views/external_users/claims/_date_attended_fields_condenced.html.haml [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+        = validation_error_message(@error_presenter, "#{error_prefix}_date")
+
+    .column-one-third.fee-date-to{class: error_class?(@error_presenter, "#{error_prefix}_date_to")}
+      %span.auto-width{id: "#{error_prefix}_date_to"}
+        = f.label :date, t('.date_to')
+        = f.gov_uk_date_field :date_to
+        = validation_error_message(@error_presenter, "#{error_prefix}_date_to")
+
+    .column-one-third.fee-date-remove
+      = link_to_remove_association t('.remove'), f
```
Comment (lostie) [inline app/views/external_users/claims/_date_attended_fields_condenced.html.haml]: Related with this [comment](https://github.com/ministryofjustice/Claim-for-Crown-Court-Defence/pull/2362/files#r187652882), seems like there's already some locales define for it?

File: features/page_objects/sections/expense_section.rb [hunk 16]
```diff
@@ -12,7 +12,7 @@ class ExpenseSection < SitePrism::Section
   element :amount, "#claim_expenses_attributes_0_amount"
   element :vat_amount, "#claim_expenses_attributes_0_vat_amount"
 
-  section :expense_date, "fieldset#expense_1_date" do
+  section :expense_date, "div#expense_1_date" do
```
Comment (lostie) [inline features/page_objects/sections/expense_section.rb]: **[not a blocker]** is there any point in leaving changing the DOM element every time given this section is already identified by and ID anyway? I would consider removing the DOM element all-together and leave the section as  `"#expense_1_date"` only, preventing further changes like this one.

**Comment by lostie** [comment]
One minor thing, your commit description are a bit short and very generic (e.g. `WIP`, `Updates`) which makes it a bit harder to get the gist of what its actually trying to achieve.

💭 Consider re-wording/squashing them into some more descriptive commits.

File: app/assets/javascripts/application.js (hunk 9 above)
Comment (cleargif) [reply app/assets/javascripts/application.js:91]: Yes - this is correct

**Comment by cleargif** [comment]
@lostie @jsugarman 

I have redone the PR and addressed the issues raised.

It has also been rebased with little conflict.

Please review.

**Comment by lostie** [comment]
LGTM 👍

**Review by lostie — APPROVED** [review]

**Pull request merged**